### PR TITLE
[EIP-1271] `setSafeSettings` + `getOffChainSignature`

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

Partially resolves [#969](https://github.com/safe-global/web-core/issues/1206)

## How this PR fixes it

A new `setSafeSettings` method has been added to the SDK, built upon a new `safe_setSettings` RPC call, which accepts the following payload:

```ts
{
  offChainSigning: boolean
}
```

It returns the newly set `SafeSettings` in the UI.

The above toggles a flag for enabling/disabling off-chain signing inside the `AppFrame`. The signing response types have been adjusted to a union as they could return the following:

```ts
{
  safeTxHash: string // on-chain
} |
{
  messageHash: string // off-chain
}
```

The reasoning behind the addition of a new RPC call can be found [here](https://www.notion.so/gnosis-safe/dApp-off-chain-capability-65a48c6a17874a2fb3dc1d79877a6bcf).

### CODE REVIEW

**Comment by schmanu** [comment]
I have read the CLA Document and I hereby sign the CLA

File: examples/wagmi/package.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "version": "0.1.3",
   "private": true,
   "dependencies": {
-    "@safe-global/safe-apps-provider": "^0.15.2",
+    "@safe-global/safe-apps-provider": "^0.16.0",
```
Comment (mikhailxyz) [inline examples/wagmi/package.json:6]: the changeset library should handle all the version changes afterwards, no need to update them manually

File: .changeset/breezy-cooks-own.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+'@gnosis.pm/safe-apps-provider': minor
+'@gnosis.pm/safe-apps-sdk': minor
+'@gnosis.pm/safe-apps-test-app': minor
+---
+
+adds new `setSafeSettings` functionality
```
Comment (mikhailxyz) [inline .changeset/breezy-cooks-own.md:7]: its incomplete innit?

File: packages/safe-apps-provider/package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@safe-global/safe-apps-provider",
-  "version": "0.15.2",
+  "version": "0.16.0",
```
Comment (mikhailxyz) [inline packages/safe-apps-provider/package.json:3]: the changeset library should handle all the version changes afterwards, no need to update them manually

File: packages/safe-apps-provider/package.json [hunk 4]
```diff
@@ -25,7 +25,7 @@
   },
   "homepage": "https://github.com/safe-global/safe-apps-sdk#readme",
   "dependencies": {
-    "@safe-global/safe-apps-sdk": "7.9.0",
+    "@safe-global/safe-apps-sdk": "7.10.0",
```
Comment (mikhailxyz) [inline packages/safe-apps-provider/package.json:28]: same here about version changes

File: packages/safe-apps-react-sdk/package.json [hunk 5]
```diff
@@ -13,7 +13,7 @@
   "license": "MIT",
   "author": "Safe (https://safe.global)",
   "dependencies": {
-    "@safe-global/safe-apps-sdk": "7.9.0"
+    "@safe-global/safe-apps-sdk": "7.10.0"
```
Comment (mikhailxyz) [inline packages/safe-apps-react-sdk/package.json:16]: same here about versions

File: packages/safe-apps-sdk/README.md [hunk 6]
```diff
@@ -171,6 +171,48 @@ try {
 
 ### Signing messages
 
+Because the Safe is a smart contract wallet, it doesn't have a private key that the wallet can use to sign messages. Instead, we have a library to sign messages, and the validation logic follows [EIP-1271 - Standard Signature Validation Method for Contracts](https://eips.ethereum.org/EIPS/eip-1271).
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/README.md:174]: This seems to be duplicated from the on-chain signing section

File: packages/safe-apps-sdk/package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@safe-global/safe-apps-sdk",
-  "version": "7.9.0",
+  "version": "7.10.0",
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/package.json:3]: same here regarding version

File: packages/safe-apps-sdk/src/safe/safe.test.ts [hunk 8]
```diff
@@ -377,6 +377,17 @@ describe('Safe Apps SDK safe methods', () => {
     });
   });
 
+  describe('SDK.safe.getOffChainSignature', () => {
+    test('Should request the off-chain signature from the interface', () => {
+      sdkInstance.safe.getOffChainSignature('0x123');
+
+      expect(postMessageSpy).toHaveBeenCalledWith(
+        expect.objectContaining({ method: Methods.getOffChainSignature }),
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/src/safe/safe.test.ts:385]: we should also check it passes the message hash right?

**Comment by iamacook** [comment]
I have read the CLA Document and I hereby sign the CLA

File: packages/safe-apps-sdk/dist/package.json [hunk 9]
```diff
@@ -4,7 +4,7 @@
     "description": "SDK developed to integrate third-party apps with Safe app.",
     "main": "dist/src/index.js",
     "typings": "dist/src/index.d.ts",
-    "_files": [
+    "files": [
```
Comment (mikhailxyz) [inline packages/safe-apps-sdk/dist/package.json:7]: this has to start with underscore on npm will go craaazy

File: packages/safe-apps-test-app/src/App.tsx [hunk 10]
```diff
@@ -42,15 +42,16 @@ const App = (): React.ReactElement => {
       });
 
     async function loadSafeInfo() {
-      const safuInfo = await SDK.safe.getInfo()
```
Comment (mikhailxyz) [inline packages/safe-apps-test-app/src/App.tsx:45]: funds aren't safu anymore?

**Review by mikhailxyz — APPROVED** [review]
splendid job mush

**Pull request merged**